### PR TITLE
Added taxonomy validation for infrastructure json object

### PR DESCRIPTION
**Pull request by YingchaoDu**
Fixes issue #1198. 
Signed-off-by: Yingchao Du <yingchao.du@ing.com>

### CODE REVIEW

File: pkg/infrastructure/attribute_manager.go [hunk 1]
```diff
@@ -46,12 +51,27 @@ func readInfrastructure() (infraattributes.Infrastructure, error) {
 	if err != nil {
 		return attributes, err
 	}
+	if err := validateStructure(content, ValidationPath); err != nil {
```
Comment (shlomitk1) [inline pkg/infrastructure/attribute_manager.go:54]: Shouldn't it be `err =`  instead of `err :=` because err is already defined?

Comment (roee88) [reply pkg/infrastructure/attribute_manager.go:54]: `err` here is scoped to the if statement

File: samples/taxonomy/infrastructure/attribute.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+definitions:
+  Attribute:
+    enum:
+      - storage-cost
+      - bandwidth
```
Comment (shlomitk1) [inline samples/taxonomy/infrastructure/attribute.yaml:5]: new line at the end of the file

File: samples/taxonomy/infrastructure/units.yaml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+definitions:
+  Units:
+    enum:
+      - US-dollar
+      - GB
+      - GB/sec
```
Comment (shlomitk1) [inline samples/taxonomy/infrastructure/units.yaml:6]: new line at the end of the file

File: samples/taxonomy/infrastructure/units.yaml [hunk 4]
```diff
@@ -0,0 +1,6 @@
+definitions:
+  Units:
+    enum:
+      - US-dollar
```
Comment (shlomitk1) [inline samples/taxonomy/infrastructure/units.yaml:4]: US Dollar (to be consistent with an example)

**Review by shlomitk1 — COMMENTED** [review]
Add tests for validating a correct infrastructure object and an invalid infrastructure object

File: samples/taxonomy/infrastructure/units.yaml (hunk 3 above)
Comment (shlomitk1) [reply samples/taxonomy/infrastructure/units.yaml]: Replace GB/sec with GB/s
Add also Mbp/s, Gbp/s, Tbp/s, MB/s, TB/s, MB, KB, TB

File: pkg/infrastructure/attribute_manager.go [hunk 5]
```diff
@@ -47,12 +52,27 @@ func readInfrastructure() (infraattributes.Infrastructure, error) {
...
+		return attributes, err
+	}
 	if err = json.Unmarshal(content, &attributes); err != nil {
 		return attributes, errors.Wrap(err, "could not parse infrastructure json")
 	}
 	return attributes, nil
 }
 
+func validateStructure(bytes []byte, taxonomySchema string) error {
+	allErrs, err := validate.TaxonomyCheck(bytes, ValidationPath)
```
Comment (shlomitk1) [inline pkg/infrastructure/attribute_manager.go:65]: Use taxonomySchema instead of ValidationPath in the function call

**Review by shlomitk1 — APPROVED** [review]

File: samples/taxonomy/infrastructure/units.yaml [hunk 6]
```diff
@@ -0,0 +1,14 @@
+definitions:
+  Units:
```
Comment (simanadler) [inline samples/taxonomy/infrastructure/units.yaml:2]: Do we want all the different types of units in one enum?  Or should there be separate ones for cost, size, etc?

File: pkg/infrastructure/attribute_manager.go [hunk 7]
```diff
@@ -47,12 +52,27 @@ func readInfrastructure() (infraattributes.Infrastructure, error) {
...
+	if err = validateStructure(content, ValidationPath); err != nil {
+		return attributes, err
+	}
 	if err = json.Unmarshal(content, &attributes); err != nil {
 		return attributes, errors.Wrap(err, "could not parse infrastructure json")
 	}
 	return attributes, nil
 }
 
+func validateStructure(bytes []byte, taxonomySchema string) error {
```
Comment (simanadler) [inline pkg/infrastructure/attribute_manager.go:64]: Should this be a general helper function provided by the taxonomy, or is it truly specific to the attribute manager?

Comment (shlomitk1) [reply pkg/infrastructure/attribute_manager.go:64]: It's just a wrapper for the function in `validate ` package

File: samples/taxonomy/infrastructure/units.yaml (hunk 6 above)
Comment (shlomitk1) [reply samples/taxonomy/infrastructure/units.yaml:2]: Will be handled in https://github.com/fybrik/fybrik/issues/1331

**Pull request merged**